rust/phoenix-channel/src/lib.rs

```diff
@@ -178,7 +260,58 @@ where
             reconnect_backoff,
             url: url.clone(),
             user_agent: user_agent.clone(),
-            state: State::connect(url, user_agent),
+            state: State::connect(url, user_agent, ()),
+            socket_protector: (),
+            waker: None,
+            pending_messages: Default::default(),
+            _phantom: PhantomData,
+            heartbeat: Heartbeat::new(
+                heartbeat::INTERVAL,
+                heartbeat::TIMEOUT,
+                next_request_id.clone(),
+            ),
+            next_request_id,
+            pending_join_requests: Default::default(),
+            login,
+            init_req,
+        }
+    }
+}
+
+impl<TInitReq, TInboundMsg, TOutboundRes, SProtect>
+    PhoenixChannel<TInitReq, TInboundMsg, TOutboundRes, SProtect>
+where
+    TInitReq: Serialize + Clone,
+    TInboundMsg: DeserializeOwned,
+    TOutboundRes: DeserializeOwned,
+    SProtect: SocketProtect<tokio::net::TcpSocket> + Clone + Send + 'static,
+{
+    /// Creates a new [PhoenixChannel] to the given endpoint.
+    ///
+    /// This version of the function accept a Protect struct that will be used to protect the socket and prevent routing loops.
+    ///
+    /// The provided URL must contain a host.
+    /// Additionally, you must already provide any query parameters required for authentication.
+    ///
+    /// Once the connection is established,
+    pub fn connect_with_protect(
+        url: Secret<LoginUrl>,
+        user_agent: String,
+        login: &'static str,
+        init_req: TInitReq,
+        reconnect_backoff: ExponentialBackoff,
+        socket_protector: SProtect,
```

Comment:
Instead of making this specifically about `protect`, the more general abstraction would be a socket factory which you can model as a `Fn(Url) -> tokio::net::TcpSocket`. Then, the concern of these needing to be protected can all be managed higher-up.